kore/src/Kore/Step.hs

```diff
@@ -49,119 +57,152 @@ import qualified Kore.Step.Simplification.Pattern as Pattern
+
+limitedExecutionStrategy :: Limit Natural -> [Strategy Prim]
+limitedExecutionStrategy depthLimit =
+    Limit.takeWithin depthLimit (toList executionStrategy)
+
+{- The primitive transitions for the symbolic execution strategy.
+-}
+data Prim
+    = Begin
+    | Simplify
+    | ApplyRewrites
     deriving (Show)
 
--- | Apply the rewrite.
-rewrite :: rewrite -> Prim rewrite
-rewrite = Rewrite
-
--- | Apply builtin simplification rewrites and evaluate functions.
-simplify :: Prim rewrite
-simplify = Simplify
-
-{- | A single-step strategy which applies the given rewrite rule.
-
-If the rewrite is successful, the built-in simplification rules and function
-evaluator are applied (see 'Pattern.simplify' for details).
-
- -}
-rewriteStep :: rewrite -> Strategy (Prim rewrite)
-rewriteStep a =
-    Strategy.sequence [Strategy.apply (rewrite a), Strategy.apply simplify]
+{- The two modes of symbolic execution. Each mode determines the way
+    rewrite rules are applied during a rewrite step.
+-}
+data ExecutionMode = All | Any
+    deriving (Show)
 
 {- | @TransitionRule@ is the general type of transition rules over 'Prim'.
  -}
 type TransitionRule monad rule state =
-    Prim rule -> state -> Strategy.TransitionT rule monad state
+    Prim -> state -> Strategy.TransitionT rule monad state
 
 {- | Transition rule for primitive strategies in 'Prim'.
-
-@transitionRule@ is intended to be partially applied and passed to
-'Strategy.runStrategy'.
  -}
 transitionRule
-    ::  forall simplifier
-    .   MonadSimplify simplifier
-    =>  TransitionRule simplifier
+    :: forall simplifier
+    .  MonadSimplify simplifier
+    => [[RewriteRule RewritingVariableName]]
+    -> ExecutionMode
+    -> TransitionRule simplifier
             (RewriteRule RewritingVariableName)
-            (Pattern RewritingVariableName)
-transitionRule =
-    \case
-        Simplify -> transitionSimplify
-        Rewrite a -> transitionRewrite a
+            (ProgramState (Pattern RewritingVariableName))
+transitionRule rewriteGroups = transitionRuleWorker
   where
+    transitionRuleWorker _ Begin (Rewritten a) = pure $ Start a
+    transitionRuleWorker _ Begin (Remaining _) = empty
+    transitionRuleWorker _ Begin state = pure state
+
+    transitionRuleWorker _ Simplify (Rewritten patt) =
+        Rewritten <$> transitionSimplify patt
+    transitionRuleWorker _ Simplify (Remaining patt) =
+        Remaining <$> transitionSimplify patt
+    transitionRuleWorker _ Simplify (Start patt) =
+        Start <$> transitionSimplify patt
+
+    transitionRuleWorker All ApplyRewrites (Remaining patt) =
+        transitionAllRewrite patt
+    transitionRuleWorker All ApplyRewrites (Start patt) =
+        transitionAllRewrite patt
+    transitionRuleWorker Any ApplyRewrites (Remaining patt) =
+        transitionAnyRewrite patt
+    transitionRuleWorker Any ApplyRewrites (Start patt) =
+        transitionAnyRewrite patt
+    transitionRuleWorker _ ApplyRewrites state = pure state
```

Comment:
We can reduce duplication with a simple helper function. Explicitly matching on all the `ProgramState` constructors makes the logic clearer and protects us if we ever add constructors to `ProgramState`.

```suggestion
    transitionRuleWorker mode ApplyRewrites (Remaining patt) =
        transitionRewrite mode patt
    transitionRuleWorker mode ApplyRewrites (Start patt) =
        transitionRewrite mode patt
    transitionRuleWorker _ ApplyRewrites state@(Rewritten patt) = pure state
        
    transitionRewrite All = transitionAllRewrite
    transitionRewrite Any = transitionAnyRewrite
```